### PR TITLE
implement a global deadline when running testcode in the local arangosh

### PR DESCRIPTION
**Pull request by dothebart**
this enables us to have a timeout for testfiles that we eval' into the testing.js arangosh by `runInLocalArangosh`.

### CODE REVIEW

**Comment by dothebart** [comment]
The results of this can be observed by specifying a small timeout:
`./scripts/unittest shell_client --oneTestTimeout 10`
```
2020-02-20T12:26:59.397Z [============] runInLocalArangosh: Trying tests/js/common/shell/aql-index-usage-rocksdb.js ... 

2020-02-20T12:26:59.400Z [------------] 1 tests from IndexUsageSuite (setUpAll: 0ms)
2020-02-20T12:26:59.400Z [ RUN        ] testIndexUsage
2020-02-20T12:27:09.404Z [   FAILED   ] testIndexUsage (setUp: 101ms, test: 9902ms, tearDown: 1ms)
ArangoError: Timeout for remote!
    at ArangoCollection.getIndexes.ArangoCollection.indexes (/home/willi/src/devel4/js/client/modules/@arangodb/arango-collection.js:671:50)
    at Object.testIndexUsage (tests/js/common/shell/aql-index-usage-rocksdb.js:101:34)
    at Object.run (/home/willi/src/devel4/js/common/modules/jsunity/jsunity.js:549:27)
    at Object.Run [as run] (/home/willi/src/devel4/js/common/modules/jsunity.js:275:24)
    at tests/js/common/shell/aql-index-usage-rocksdb.js:128:9
    at tests/js/common/shell/aql-index-usage-rocksdb.js:131:3
    at RunTest (/home/willi/src/devel4/js/common/modules/jsunity.js:389:12)
    at testFunc (eval at runInLocalArangosh (/home/willi/src/devel4/js/client/modules/@arangodb/test-utils.js:889:3), <anonymous>:4:9)
    at runInLocalArangosh (/home/willi/src/devel4/js/client/modules/@arangodb/test-utils.js:893:18)
    at Object.performTests (/home/willi/src/devel4/js/client/modules/@arangodb/test-utils.js:233:21) - test failed
ArangoError: Timeout for remote!
    at ArangoCollection.drop (/home/willi/src/devel4/js/client/modules/@arangodb/arango-collection.js:510:48)
    at Proxy.ArangoDatabase._drop (/home/willi/src/devel4/js/client/modules/@arangodb/arango-database.js:442:29)
    at Object.tearDown (tests/js/common/shell/aql-index-usage-rocksdb.js:57:10)
    at /home/willi/src/devel4/js/common/modules/jsunity/jsunity.js:483:16
    at Object.run (/home/willi/src/devel4/js/common/modules/jsunity/jsunity.js:554:19)
    at Object.Run [as run] (/home/willi/src/devel4/js/common/modules/jsunity.js:275:24)
    at tests/js/common/shell/aql-index-usage-rocksdb.js:128:9
    at tests/js/common/shell/aql-index-usage-rocksdb.js:131:3
    at RunTest (/home/willi/src/devel4/js/common/modules/jsunity.js:389:12)
    at testFunc (eval at runInLocalArangosh (/home/willi/src/devel4/js/client/modules/@arangodb/test-utils.js:889:3), <anonymous>:4:9) - tearDown failed
2020-02-20T12:27:09.404Z [------------] 1 tests from IndexUsageSuite ran (tearDownAll: 1ms)
2020-02-20T12:27:09.404Z [   PASSED   ] 0 tests.
2020-02-20T12:27:09.404Z [   FAILED   ] 1 tests.
2020-02-20T12:27:09.404Z [============] Ran: 1 tests (0 passed, 1 failed) (10004ms total)

oops! Skipping remaining tests, server is gone.
Thu Feb 20 2020 13:27:09 GMT+0100 (Central European Standard Time) Shutting down...
```

File: lib/V8/v8-utils.cpp [hunk 1]
```diff
@@ -202,6 +202,86 @@ static void CreateErrorObject(v8::Isolate* isolate, int errorNumber,
...
+    return timeout;
+  }
+  auto now = TRI_microtime();
+  auto delta = now - when;
+  if (delta > timeout) {
+    return timeout;
+  }
+  return delta;
+}
+#define ARANGODB_SHELL_V8CLIENT_CONNECTION_H 1
```
Comment (ObiWahn) [inline lib/V8/v8-utils.cpp]: Should this be a `#ifdef` ... `#endif`?

File: lib/V8/v8-utils.cpp [hunk 2]
```diff
@@ -202,6 +202,86 @@ static void CreateErrorObject(v8::Isolate* isolate, int errorNumber,
...
+  if (now < when) {
+    return false;
+  }
+
+  auto isolate = args.GetIsolate();
+  TRI_CreateErrorObject(isolate, TRI_ERROR_DISABLED, "Timeout for remote!", true);
+  return true;
+}
+
+double getMaxTimeoutConnectionToBeDead(double timeout) {
```
Comment (ObiWahn) [inline lib/V8/v8-utils.cpp]: Do we really need 2 versions of the functions?

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8668/

File: lib/V8/v8-utils.cpp (hunk 1 above)
Comment (dothebart) [reply lib/V8/v8-utils.cpp]: https://en.cppreference.com/w/cpp/feature_test - replaced by `#if __cpp_lib_chrono`

File: lib/V8/v8-utils.cpp (hunk 2 above)
Comment (dothebart) [reply lib/V8/v8-utils.cpp]: as long as js_download works with `double` seconds as a timeout, we need this other function. 
Removed Polimorphic naming, as they do different things.

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8728/

File: lib/V8/v8-utils.cpp (hunk 1 above)
Comment (ObiWahn) [reply lib/V8/v8-utils.cpp]: That is a c++20 feature! Does it work for all our compilers?

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8748/

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8843/

File: lib/V8/v8-utils.cpp [hunk 3]
```diff
@@ -21,6 +21,9 @@
 /// @author Dr. Frank Celler
 ////////////////////////////////////////////////////////////////////////////////
 
+#include <chrono>
+#define arangodb_have_cpp_lib_chrono 1
```
Comment (jsteemann) [inline lib/V8/v8-utils.cpp]: we should remove this define, and just use `#include <chrono>` in the header where it is now used.

File: lib/V8/v8-utils.cpp [hunk 4]
```diff
@@ -202,6 +204,85 @@ static void CreateErrorObject(v8::Isolate* isolate, int errorNumber,
   }
 }
 
+////////////////////////////////////////////////////////////////////////////////
+/// @brief set a point in time after which we will abort external connection
+////////////////////////////////////////////////////////////////////////////////
+static std::atomic<double> connectionToBeDeadAt(0.0);
```
Comment (jsteemann) [inline lib/V8/v8-utils.cpp]: `dead` is a wrong adjective here, as the connection is not dead.
The variable name should rather be `connectionExpirationStamp` or something like this.

File: lib/V8/v8-utils.h [hunk 5]
```diff
@@ -102,6 +102,18 @@ static T* TRI_UnwrapClass(v8::Handle<v8::Object> obj, int32_t type,
...
+////////////////////////////////////////////////////////////////////////////////
+/// @brief set a point in time after which we will abort external connection
+////////////////////////////////////////////////////////////////////////////////
+void setConnectionToBeDeadInMS(uint64_t timeout);
+bool isConnectionToBeDead(v8::FunctionCallbackInfo<v8::Value> const& args);
+double getMaxTimeoutConnectionToBeDeadS(double timeoutSeconds);
+// You will only see this function if you #include <chrono>
+#ifdef arangodb_have_cpp_lib_chrono
+std::chrono::milliseconds getMaxTimeoutConnectionToBeDead(std::chrono::milliseconds timeout);
+#endif
```
Comment (jsteemann) [inline lib/V8/v8-utils.h]: I don't see any need for this. Why not just include `chrono` here?

File: lib/V8/v8-utils.cpp (hunk 4 above)
Comment (jsteemann) [reply lib/V8/v8-utils.cpp]: Additionally, as all of this seems to be used from within the arangosh only, can we move the code from `lib` into the `arangosh` folder?

Comment (dothebart) [reply lib/V8/v8-utils.cpp]: the simple http client is used on both sides :/

File: lib/V8/v8-utils.cpp [hunk 6]
```diff
@@ -202,6 +202,85 @@ static void CreateErrorObject(v8::Isolate* isolate, int errorNumber,
...
+  }
+
+  auto isolate = args.GetIsolate();
+  TRI_CreateErrorObject(isolate, TRI_ERROR_DISABLED, "Timeout for remote!", true);
+  return true;
+}
+
+double getMaxTimeoutConnectionToBeDeadS(double timeoutSeconds) {
+  auto when = connectionToBeDeadAt.load();
+  if (when == 0.0) {
```
Comment (dothebart) [inline lib/V8/v8-utils.cpp]: < 0.0001

File: lib/V8/v8-utils.cpp [hunk 7]
```diff
@@ -202,6 +202,85 @@ static void CreateErrorObject(v8::Isolate* isolate, int errorNumber,
...
+    return timeoutSeconds;
+  }
+  return delta;
+}
+
+std::chrono::milliseconds getMaxTimeoutConnectionToBeDead(std::chrono::milliseconds timeout) {
+  using namespace std::chrono;
+
+  double epochDoubleWhen = connectionToBeDeadAt.load();
+  if (epochDoubleWhen == 0.0) {
```
Comment (dothebart) [inline lib/V8/v8-utils.cpp]: < 0.0001

**Comment by dothebart** [comment]
Rename "deadline"

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8874/

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8900/

File: lib/V8/v8-utils.cpp [hunk 8]
```diff
@@ -202,6 +202,85 @@ static void CreateErrorObject(v8::Isolate* isolate, int errorNumber,
   }
 }
 
+////////////////////////////////////////////////////////////////////////////////
+/// @brief set a point in time after which we will abort external connection
+////////////////////////////////////////////////////////////////////////////////
+static std::atomic<double> executionDeadline(0.0);
```
Comment (ObiWahn) [inline lib/V8/v8-utils.cpp]: Is there just a single thread?

File: lib/V8/v8-utils.cpp [hunk 9]
```diff
@@ -202,6 +202,85 @@ static void CreateErrorObject(v8::Isolate* isolate, int errorNumber,
...
+  if (now < when) {
+    return false;
+  }
+
+  auto isolate = args.GetIsolate();
+  TRI_CreateErrorObject(isolate, TRI_ERROR_DISABLED, "Deadline for remote reached!", true);
+  return true;
+}
+
+double correctTimeoutToExecutionDeadlineS(double timeoutSeconds) {
```
Comment (ObiWahn) [inline lib/V8/v8-utils.cpp]: Does this function do the same as the function below? If so is there no way to express one with help of the other?

File: lib/V8/v8-utils.cpp (hunk 8 above)
Comment (dothebart) [reply lib/V8/v8-utils.cpp]: in arangosh - yes. however, it should be correct for arangod.

File: lib/V8/v8-utils.cpp (hunk 9 above)
Comment (dothebart) [reply lib/V8/v8-utils.cpp]: this version can probably be erased along with the simple http client - thus I don't want to spend more time on this.

File: lib/V8/v8-utils.cpp (hunk 8 above)
Comment (ObiWahn) [reply lib/V8/v8-utils.cpp]: I am not sure. Why should it be correct to have a single variable for potentially multiple contexts?

**Comment by dothebart** [comment]
Since each of you @ObiWahn @KVS85 @jsteemann  raised concerns about this functionality also being available there in the server, I modified it to be available in clients only. 

http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8911/

**Comment by ObiWahn** [comment]
You use `setExecutionDeadlineInMS` only in the function that is available to the client. But it is not obvious at all that people are not allowed to use this function in the server. You need at least some comment or better move the code altogether to arangosh.

File: lib/V8/v8-utils.cpp [hunk 10]
```diff
@@ -202,6 +202,85 @@ static void CreateErrorObject(v8::Isolate* isolate, int errorNumber,
   }
 }
 
+////////////////////////////////////////////////////////////////////////////////
+/// @brief set a point in time after which we will abort external connection
+////////////////////////////////////////////////////////////////////////////////
+static double executionDeadline = 0.0;
```
Comment (ObiWahn) [inline lib/V8/v8-utils.cpp]: double

File: lib/V8/v8-utils.cpp [hunk 11]
```diff
@@ -202,6 +202,85 @@ static void CreateErrorObject(v8::Isolate* isolate, int errorNumber,
...
+void setExecutionDeadlineInMS(uint64_t timeout) {
```
Comment (ObiWahn) [inline lib/V8/v8-utils.cpp]: uint64_t

File: lib/V8/v8-utils.cpp [hunk 12]
```diff
@@ -202,6 +202,85 @@ static void CreateErrorObject(v8::Isolate* isolate, int errorNumber,
...
+  }
+  auto now = TRI_microtime();
+  auto delta = when - now;
+  if (delta > timeoutSeconds) {
+    return timeoutSeconds;
+  }
+  return delta;
+}
+
+std::chrono::milliseconds correctTimeoutToExecutionDeadline(std::chrono::milliseconds timeout) {
```
Comment (ObiWahn) [inline lib/V8/v8-utils.cpp]: std::chrono::milliseconds

**Comment by dothebart** [comment]
Again - we may fix the usage of chrono once we do this for js_download. its out of scope for this PR.

**Comment by ObiWahn** [comment]
Just use chrono for your new code and cast to others as required.

**Comment by dothebart** [comment]
I (we) did, on all places that already use it. `js_download` will most probably completely be replaced soon to get rid of that http implementation, at that time the implementation can simply be erased.

File: lib/V8/v8-utils.cpp [hunk 13]
```diff
@@ -214,6 +214,7 @@ void setExecutionDeadlineInMS(uint64_t timeout) {
   }
 }
 
+// arangosh only: set a deadline
```
Comment (ObiWahn) [inline lib/V8/v8-utils.cpp]: `setExecutionDeadlineInMS` should have the information.

**Comment by dothebart** [comment]
tests green.

File: lib/V8/v8-utils.cpp (hunk 11 above)
Comment (dothebart) [reply lib/V8/v8-utils.cpp]: removed function.

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/9010/

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/9012/

**Comment by dothebart** [comment]
tests all blue.

File: lib/V8/v8-deadline.cpp [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+#include "V8/v8-conv.h"
+#include "V8/v8-globals.h"
+
+////////////////////////////////////////////////////////////////////////////////
+/// @brief set a point in time after which we will abort external connection
+////////////////////////////////////////////////////////////////////////////////
+static double executionDeadline = 0.0;
+
+// arangosh only: set a deadline
+static void JS_SetExecutionDeatlineTo(v8::FunctionCallbackInfo<v8::Value> const& args) {
```
Comment (jsteemann) [inline lib/V8/v8-deadline.cpp]:
```suggestion
static void JS_SetExecutionDeadlineTo(v8::FunctionCallbackInfo<v8::Value> const& args) {
```

File: lib/V8/v8-deadline.cpp [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+/// @author Wilfried Goesgens
+////////////////////////////////////////////////////////////////////////////////
+#include "V8/v8-deadline.h"
+#include "Basics/system-functions.h"
+#include "v8-utils.h"
+#include "V8/v8-conv.h"
+#include "V8/v8-globals.h"
+
+////////////////////////////////////////////////////////////////////////////////
+/// @brief set a point in time after which we will abort external connection
```
Comment (jsteemann) [inline lib/V8/v8-deadline.cpp]:
```suggestion
/// @brief set a point in time after which we will abort certain operations
```

File: lib/V8/v8-deadline.cpp [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+  if (when < 0.00001) {
+    return false;
+  }
+  auto now = TRI_microtime();
+  if (now < when) {
+    return false;
+  }
+
+  auto isolate = args.GetIsolate();
+  TRI_CreateErrorObject(isolate, TRI_ERROR_DISABLED, "Deadline for remote reached!", true);
```
Comment (jsteemann) [inline lib/V8/v8-deadline.cpp]:
```suggestion
  TRI_CreateErrorObject(isolate, TRI_ERROR_DISABLED, "Execution deadline reached!", true);
```

File: lib/V8/v8-deadline.cpp [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+void TRI_InitV8Deadline(v8::Isolate* isolate,
+                        v8::Handle<v8::Context>,
+                        std::string const& startupPath,
+                        std::string const& modules)
+{
+  TRI_AddGlobalFunctionVocbase(isolate,
+                               TRI_V8_ASCII_STRING(isolate, "SYS_COMMUNICATE_SLEEP_DEADLINE"),
+                               JS_SetExecutionDeatlineTo);
```
Comment (jsteemann) [inline lib/V8/v8-deadline.cpp]:
```suggestion
                               JS_SetExecutionDeadlineTo);
```

File: lib/V8/v8-deadline.cpp [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+    return timeout;
+  }
+  return delta;
+
+}
+
+void TRI_InitV8Deadline(v8::Isolate* isolate,
+                        v8::Handle<v8::Context>,
+                        std::string const& startupPath,
+                        std::string const& modules)
```
Comment (jsteemann) [inline lib/V8/v8-deadline.cpp]: these parameters are not used here and should be removed.

File: lib/V8/v8-deadline.h [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+/// @brief set a point in time after which we will abort external connection
+////////////////////////////////////////////////////////////////////////////////
+bool isExecutionDeadlineReached(v8::FunctionCallbackInfo<v8::Value> const& args);
+double correctTimeoutToExecutionDeadlineS(double timeoutSeconds);
+std::chrono::milliseconds correctTimeoutToExecutionDeadline(std::chrono::milliseconds timeout);
+
+void TRI_InitV8Deadline(v8::Isolate* isolate,
+                        v8::Handle<v8::Context>,
+                        std::string const& startupPath,
+                        std::string const& modules);
```
Comment (jsteemann) [inline lib/V8/v8-deadline.h]: these parameters are not needed and should be removed

File: lib/V8/v8-deadline.cpp [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+    executionDeadline = 0.0;
+  } else {
+    executionDeadline = TRI_microtime() + n / 1000;
+  }
+
+  TRI_V8_RETURN_BOOL((when > 0.00001) && (now - when > 0.0) );
+  TRI_V8_TRY_CATCH_END
+}
+
+bool isExecutionDeadlineReached(v8::FunctionCallbackInfo<v8::Value> const& args) {
```
Comment (jsteemann) [inline lib/V8/v8-deadline.cpp]: I was a bit surprised why this function would get `args` from multiple callers without knowing what the args mean.
But it seems like it does not need _args_ but the _isolate_. So we shouldn't pass args, but the isolate to this function.

File: lib/V8/v8-deadline.h [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+#define ARANGODB_V8_V8__DEADLINE_H 1
+
+#include <cstdint>
+#include <chrono>
+#include <v8.h>
+
+////////////////////////////////////////////////////////////////////////////////
+/// @brief set a point in time after which we will abort external connection
+////////////////////////////////////////////////////////////////////////////////
+bool isExecutionDeadlineReached(v8::FunctionCallbackInfo<v8::Value> const& args);
```
Comment (jsteemann) [inline lib/V8/v8-deadline.h]: I was a bit surprised why this function would get `args` from multiple callers without knowing what the args mean.
But it seems like it does not need _args_ but the _isolate_. So we shouldn't pass args, but the isolate to this function.

File: lib/V8/v8-deadline.h [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Wilfried Goesgens
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef ARANGODB_V8_V8__DEADLINE_H
+#define ARANGODB_V8_V8__DEADLINE_H 1
+
+#include <cstdint>
```
Comment (jsteemann) [inline lib/V8/v8-deadline.h]: Not needed for this header file
```suggestion
```

File: lib/V8/v8-no-deadline.cpp [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+std::chrono::milliseconds correctTimeoutToExecutionDeadline(std::chrono::milliseconds timeout) {
+  return timeout;
+}
+
+void TRI_InitV8Deadline(v8::Isolate* isolate,
+                        v8::Handle<v8::Context>,
+                        std::string const& startupPath,
+                        std::string const& modules)
```
Comment (jsteemann) [inline lib/V8/v8-no-deadline.cpp]: Parameters not needed here.

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/9022/

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/9026/

**Pull request merged**